/**:
  ros__parameters:
    # --- Node Settings ---
    publish_global_tf: true # Publish the map -> odom transform.
    publish_smoothed_path: true # Publish the full smoothed path for visualization.
    publish_velocity: true # Publish the estimated velocity (of DVL frame in map frame).
    publish_imu_bias: true # Publish the estimated IMU biases (accel + gyro).

    # These are computationally expensive to calculate, so disable if not needed.
    publish_pose_cov: true # Publish the pose covariance along with the pose.
    publish_velocity_cov: false # Publish the velocity covariance along with the velocity.
    publish_imu_bias_cov: false # Publish the IMU bias covariance along with the bias.

    # --- GTSAM Smoother Settings ---
    smoother_lag: 20.0 # The time window (in seconds) of past states to smooth over.
    relinearize_threshold: 0.1 # Lower values are more accurate but slower.
    relinearize_skip: 1 # Number of updates to skip between relinearizations.

    # --- ROS Topics and Frames ---
    imu_topic: "imu/data"
    gps_odom_topic: "odometry/gps"
    depth_odom_topic: "odometry/depth"
    mag_topic: "imu/mag"
    ahrs_topic: "imu/ahrs"
    dvl_topic: "dvl/twist"
    global_odom_topic: "odometry/global"
    smoothed_path_topic: "smoothed_path"
    velocity_topic: "~/velocity" # Published at the DVL frame.
    # Topic for the IMU bias. Published as a TwistWithCovarianceStamped,
    # where 'linear' holds accelerometer bias and 'angular' holds gyroscope bias.
    imu_bias_topic: "~/imu_bias"

    # --- Sensor Noise Models ---
    # For each sensor, `use_parameter_covariance: true` will force the node to
    # ignore the covariance in the sensor message and use the sigma value(s)
    # specified in this file instead.
    #
    # Robust Kernels:
    # - "None": Standard Least Squares (Gaussian). Sensitive to outliers.
    # - "Huber": Linear penalty for large errors. Good general robustness. (k=1.345 standard)
    # - "Tukey": Ignores large errors completely. Best for sensors with distinct 'bad' readings. (k=4.685 standard)

    imu:
      use_parameter_frame: false
      use_parameter_covariance: false
      parameter_covariance:
        accel_noise_sigma: 0.0078 # [m/s^2] Accelerometer measurement noise sigma.
        gyro_noise_sigma: 0.0012 # [rad/s] Gyroscope measurement noise sigma.
      accel_bias_rw_sigma: 1.05e-04 # [m/s^3] Sigma for accelerometer bias random walk.
      gyro_bias_rw_sigma: 3.91e-05 # [rad/s^2] Sigma for gyroscope bias random walk.
      integration_covariance: 1e-5 # A small covariance for the integration process itself.
      gravity: [0.0, 0.0, -9.8] # [m/s^2]

    gps:
      enable_gps: false
      enable_altitude: false
      use_parameter_frame: false
      use_parameter_covariance: true
      parameter_covariance:
        position_noise_sigma: 0.015 # [m] GPS position uncertainty sigma.
        altitude_noise_sigma: 0.025 # [m] GPS altitude uncertainty sigma.
      robust_kernel: "None"
      robust_k: 1.345

    depth:
      use_parameter_frame: false
      use_parameter_covariance: false
      parameter_covariance:
        position_z_noise_sigma: 0.02 # [m] Depth sensor Z-position uncertainty sigma.
      robust_kernel: "None"
      robust_k: 1.345

    mag:
      enable_mag: false
      constrain_yaw_only: true
      use_parameter_frame: false
      use_parameter_covariance: false
      parameter_covariance:
        magnetic_field_noise_sigma: 0.003 # [normalized] Magnetic field uncertainty sigma.
      # https://www.ngdc.noaa.gov/geomag/calculators/magcalc.shtml#igrfwmm
      # Reference Magnetic Field Vector (Local Earth Field)
      # Provo, Utah (X=20,843.1 nT, Y=3,894.9 nT, Z=45,873.1 nT, F=50,536.6 nT)
      reference_field: [0.41, 0.08, 0.91] # [normalized]
      robust_kernel: "None"
      robust_k: 1.345

    ahrs:
      enable_ahrs: true
      enable_roll_pitch: false
      use_parameter_frame: false
      use_parameter_covariance: false
      parameter_covariance:
        yaw_noise_sigma: 0.01745 # [rad] AHRS sensor yaw uncertainty sigma.
        roll_pitch_noise_sigma: 0.00349 # [rad] AHRS sensor roll and pitch uncertainty sigma.
      # https://www.ngdc.noaa.gov/geomag/calculators/magcalc.shtml#declination
      mag_declination_radians: 0.0
      robust_kernel: "None"
      robust_k: 1.345

    dvl:
      use_parameter_frame: false
      use_parameter_covariance: false
      parameter_covariance:
        velocity_noise_sigma: 0.02 # [m/s] DVL body-velocity uncertainty sigma.
      timeout_threshold: 1.0 # [s] Time before switching to depth-driven FGO (DVL dropout).
      robust_kernel: "None"
      robust_k: 1.345

    # --- Initial State Priors ---
    prior:
      initialization_duration: 10.0 # [s] Time duration to collect and average data before initializing.

      # If true, forces the initial state to be set by the values below,
      # ignoring any information from initial sensor messages.
      use_parameter_priors: true
      parameter_priors:
        initial_position: [0.0, 0.0, 0.0] # [m]
        initial_orientation: [0.0, 0.0, 0.0] # [rad]
        initial_velocity: [0.0, 0.0, 0.0] # [m/s] (at DVL frame)
        initial_accel_bias: [0.0, 0.0, 0.0] # [m/s^2]
        initial_gyro_bias: [0.0, 0.0, 0.0] # [rad/s]
        initial_position_sigma: 1e-3 # [m] Initial position uncertainty.
        initial_orientation_sigma: 1e-3 # [rad] Initial yaw uncertainty.
        initial_velocity_sigma: 1e-3 # [m/s] Initial velocity uncertainty.
      initial_accel_bias_sigma: 1e-3 # [m/s^2] Initial accelerometer bias uncertainty.
      initial_gyro_bias_sigma: 1e-4 # [rad/s] Initial gyroscope bias uncertainty.

    # --- Experimental Features ---
    experimental:
      # TURTLMap-based DVL preintegration implementation
      enable_dvl_preintegration: false

/**/navsat_preprocessor_node:
  ros__parameters:
    # --- Node Settings ---
    origin_pub_rate: 1.0 # [Hz] Rate at which to publish the GPS origin.
    initialization_duration: 10.0 # [s] Time duration to average GPS data before setting origin.

    simulate_dropout: false
    dropout_frequency: 0.0333 # [Hz] Frequency of the dropout cycle (1/30).
    dropout_duration: 5.0 # [s] Duration of the dropout.

    # --- ROS Topics and Frames ---
    input_topic: "gps/fix"
    odom_output_topic: "odometry/gps"
    origin_topic: "/origin"

    use_parameter_child_frame: false
    use_parameter_origin: true
    parameter_origin:
      latitude: 40.33940
      longitude: -111.90721
      altitude: 1412.0

/auv0/**:
  ros__parameters:
    prior:
      parameter_priors:
        initial_position: [0.0, 0.0, 0.0] # [m]
        initial_orientation: [0.0, 0.0, 0.0] # [rad]

/auv1/**:
  ros__parameters:
    prior:
      parameter_priors:
        initial_position: [0.0, 20.0, 0.0] # [m]
        initial_orientation: [0.0, 0.0, 0.0] # [rad]

/auv2/**:
  ros__parameters:
    prior:
      parameter_priors:
        initial_position: [0.0, 40.0, 0.0] # [m]
        initial_orientation: [0.0, 0.0, 0.0] # [rad]

/bluerov2/**:
  ros__parameters:
    ahrs_topic: "imu/data"

    imu:
      use_parameter_frame: true
      parameter_frame: "bluerov2/imu_link"

    gps:
      # enable_gps: true
      use_parameter_frame: true
      parameter_frame: "bluerov2/gps_link"

    depth:
      use_parameter_frame: true
      parameter_frame: "bluerov2/imu_link"

    ahrs:
      enable_ahrs: false
      use_parameter_frame: true
      parameter_frame: "bluerov2/imu_link"
      mag_declination_radians: 0.1876

    dvl:
      use_parameter_frame: true
      parameter_frame: "bluerov2/dvl_link"

    initialization_duration: 0.0
    use_parameter_child_frame: true
    parameter_child_frame: "bluerov2/gps_link"
    use_parameter_origin: false
